### ISSUE TITLE
Migrations: Different properties with same name in derived classes throws exception

### ISSUE THREAD

**Issue by jakimber** [aspnet/EntityFramework#5894]
as per [https://blogs.msdn.microsoft.com/dotnet/2016/06/27/entity-framework-core-1-0-0-available/#comment-91605](https://blogs.msdn.microsoft.com/dotnet/2016/06/27/entity-framework-core-1-0-0-available/#comment-91605) thanks for any time you can spend on this, I appreciate you are busy!

This may be a result of progress, but i'm a little confused and have spent a fair amount of time trawling the docs so would appreciate some guidance.  (Happy to post to StackOverflow if that is more appropriate too!)

### The issue
I have a model that I have been building and saving with RC2 for a while now, migrations created and updated just fine.  After my eager upgrade to 1.0.0 my migrations file still gets created just fine but `dotnet ef database update` fails with ` Sequence contains more than one matching element`.

It may be that work on 1.0.0 has made things more strict than before and I was merely getting lucky with the RC, but ideally I'd like a better fix to the workarounds that I have got.

### TL;DR;
I can no longer map two derived class properties that are named the same into a single TPH-style table.

### Steps to reproduce
After trying with my original project I have reproduced the error in a brand new vanilla project with `"Microsoft.EntityFrameworkCore.Tools": "1.0.0-preview2-final"` added in `"references"` and `"tools"`.

Here's a model that matches my design and causes the issue for me:
```
namespace Model
{
    // Base class that all three derived classes inherit
    public class MyBaseClass
    {
        public int Id { get; set; }
        public string BaseClassProperty { get; set; }
    }

    // interface that (in my design) only two of the three derived classes implement
    public interface IMyInterface
    {
        string MyInterfaceProperty { get; set; }
    }

    // First derived class, doesn't implement the interface (obviously has more to it in my real application)
    public class MyDerivedClass1 : MyBaseClass { }

    // and the classes that derive from base, and also implement the interface
    public class MyDerivedClass2 : MyBaseClass, IMyInterface
    {
        public string MyInterfaceProperty { get; set; }
    }

    public class MyDerivedClass3 : MyBaseClass, IMyInterface
    {
        public string MyInterfaceProperty { get; set; }
    }
}
```

So, before you say that interface properties are not supported and close my issue, I have subsequently tried the following:
### Attempt 1
Avoid the interface, add duplicate properties to the derived classes and map them like this:
```
protected override void OnModelCreating(ModelBuilder modelBuilder)
{
    modelBuilder.Entity<MyDerivedClass2>().ToTable("MyClass2").Property(c => c.MyProperty).HasColumnName("ColumnA");
    modelBuilder.Entity<MyDerivedClass3>().ToTable("MyClass3").Property(c => c.MyProperty).HasColumnName("ColumnB");
    base.OnModelCreating(modelBuilder);
}
```
This works, although having both properties mapped to `ColumnA` doesn't.  Also, it really hurts my design as i'm using that interface for some polymorphic behavior in my app.

### Attempt 2
It's this one that made me consider that maybe, just maybe, it's a tooling issue.

With just the table mapping and not the property mapping in Attempt 1, once the migrations are created, edit the mapping file:
```
protected override void OnModelCreating(ModelBuilder modelBuilder)
{
    modelBuilder.Entity<MyDerivedClass2>().ToTable("MyClass2");
    modelBuilder.Entity<MyDerivedClass3>().ToTable("MyClass3");
    base.OnModelCreating(modelBuilder);
}

// Edit the mapping file (code snipped for brevity)
protected override void Up(MigrationBuilder migrationBuilder)
{
    migrationBuilder.CreateTable(
        name: "MyBaseClasses",  <- change this to "MyBaseClass"
```

then `dotnet ef database update` works.  I then have to `ALTER TABLE MyBaseClass RENAME TO MyBaseClasses`.  The application then runs just fine, albeit a somewhat convoluted process to get there!

### Attempt 3
Change my design to use more of an object hierarchy than interface implementation.  Using a further layer in my object hierarchy (with or without the properties being `abstract`) also results in a success condition, but again, my design has to suffer somewhat to achieve this, and i'm really reticent to add layers to my object model to suit a technical issue rather than a domain one.

### Attempt 4
I also tried avoiding the naming conflict entirely in my class hierarchy with distinct property names and of course that worked, but would definitely mean I stop using EF Core for now as that just messes far too much with my zen.

I hope it's just something dumb I have missed, my search-fu is quite developed, but it's more than possible I missed something out there.  Thanks again for your help!

### Extra info
```
Exception message:
System.InvalidOperationException: Sequence contains more than one matching element    at System.Linq.Enumerable.SingleOrDefault[TSource](IEnumerable`1 source, Func`2 predicate)
    at Microsoft.EntityFrameworkCore.Migrations.MigrationsSqlGenerator.GetColumnType(String schema, String table, String name, Type clrType, Nullable`1 unicode, Nullable`1 maxLength, Boolean rowVersion, IModel model) 
    at Microsoft.EntityFrameworkCore.Migrations.MigrationsSqlGenerator.ColumnDefinition(String schema, String table, String name, Type clrType, String type, Nullable`1 unicode, Nullable`1 maxLength, Boolean rowVersion, Boolean nullable, Object defaultValue,  String defaultValueSql, String computedColumnSql, IAnnotatable annotatable, IModel model, MigrationCommandListBuilder builder)
    at Microsoft.EntityFrameworkCore.Migrations.SqliteMigrationsSqlGenerator.ColumnDefinition(String schema, String table, String name, Type clrType, String type, Nullable`1 unicode, Nullable`1 maxLength, Boolean rowVersion, Boolean nullable, Object default Value, String defaultValueSql, String computedColumnSql, IAnnotatable annotatable, IModel model, MigrationCommandListBuilder builder)
    at Microsoft.EntityFrameworkCore.Migrations.MigrationsSqlGenerator.ColumnDefinition(AddColumnOperation operation, IModel model, MigrationCommandListBuilder builder) 
    at Microsoft.EntityFrameworkCore.Migrations.MigrationsSqlGenerator.Generate(CreateTableOperation operation, IModel model, MigrationCommandListBuilder builder) 
    at Microsoft.EntityFrameworkCore.Migrations.MigrationsSqlGenerator.Generate(MigrationOperation operation, IModel model, MigrationCommandListBuilder builder)
    at Microsoft.EntityFrameworkCore.Migrations.MigrationsSqlGenerator.Generate(IReadOnlyList`1 operations, IModel model) 
    at Microsoft.EntityFrameworkCore.Migrations.Internal.Migrator.GenerateUpSql(Migration migration)
    at Microsoft.EntityFrameworkCore.Migrations.Internal.Migrator.Migrate(String targetMigration) 
    at Microsoft.EntityFrameworkCore.Design.MigrationsOperations.UpdateDatabase(String targetMigration, String contextType) 
    at Microsoft.EntityFrameworkCore.Tools.Cli.DatabaseUpdateCommand.<>c__DisplayClass0_0.<Configure>b__0() 
    at Microsoft.Extensions.CommandLineUtils.CommandLineApplication.Execute(String[] args) 
    at Microsoft.EntityFrameworkCore.Tools.Cli.Program.Main(String[] args)
 Sequence contains more than one matching element 
```

### Further technical details
EF Core version: Microsoft.NETCore.App version 1.0.0
Operating system: Windows 10
Visual Studio version: VS2015 Update 3 or VSCode 1.2.1 with omnisharp extension and the latest .NET core SDK released on the 27th.
Project details:  All .NET Core.  All upgraded to 1.0.0.  Database is sqlite.  I can probably share the actual code if you'd like a link, but the code above is a good repro and i'd rather not share it in such a public location!

**Comment by pHr34kY** [comment]
I beleive I'm getting the same issue. It all worked fine in RC2.

When I call Migrate():

```
  at System.Linq.Enumerable.SingleOrDefault[TSource](IEnumerable`1 source, Func`2 predicate)
   at Microsoft.EntityFrameworkCore.Migrations.MigrationsSqlGenerator.FindProperty(IModel model, String schema, String tableName, String columnName)
   at Microsoft.EntityFrameworkCore.Migrations.MigrationsSqlGenerator.GetColumnType(String schema, String table, String name, Type clrType, Nullable`1 unicode, Nullable`1 maxLength, Boolean rowVersion, IModel model)
   at Microsoft.EntityFrameworkCore.Migrations.MigrationsSqlGenerator.ColumnDefinition(String schema, String table, String name, Type clrType, String type, Nullable`1 unicode, Nullable`1 maxLength, Boolean rowVersion, Boolean nullable, Object defaultValue, String defaultValueSql, String computedColumnSql, IAnnotatable annotatable, IModel model, MigrationCommandListBuilder builder)
   at Microsoft.EntityFrameworkCore.Migrations.SqliteMigrationsSqlGenerator.ColumnDefinition(String schema, String table, String name, Type clrType, String type, Nullable`1 unicode, Nullable`1 maxLength, Boolean rowVersion, Boolean nullable, Object defaultValue, String defaultValueSql, String computedColumnSql, IAnnotatable annotatable, IModel model, MigrationCommandListBuilder builder)
   at Microsoft.EntityFrameworkCore.Migrations.MigrationsSqlGenerator.ColumnDefinition(AddColumnOperation operation, IModel model, MigrationCommandListBuilder builder)
   at Microsoft.EntityFrameworkCore.Migrations.MigrationsSqlGenerator.Generate(CreateTableOperation operation, IModel model, MigrationCommandListBuilder builder)
   at Microsoft.EntityFrameworkCore.Migrations.SqliteMigrationsSqlGenerator.Generate(CreateTableOperation operation, IModel model, MigrationCommandListBuilder builder)
   at Microsoft.EntityFrameworkCore.Migrations.MigrationsSqlGenerator.<>c.<.cctor>b__53_8(MigrationsSqlGenerator g, MigrationOperation o, IModel m, MigrationCommandListBuilder b)
   at Microsoft.EntityFrameworkCore.Migrations.MigrationsSqlGenerator.Generate(MigrationOperation operation, IModel model, MigrationCommandListBuilder builder)
   at Microsoft.EntityFrameworkCore.Migrations.MigrationsSqlGenerator.Generate(IReadOnlyList`1 operations, IModel model)
   at Microsoft.EntityFrameworkCore.Migrations.SqliteMigrationsSqlGenerator.Generate(IReadOnlyList`1 operations, IModel model)
   at Microsoft.EntityFrameworkCore.Migrations.Internal.Migrator.GenerateUpSql(Migration migration)
   at Microsoft.EntityFrameworkCore.Migrations.Internal.Migrator.<>c__DisplayClass13_2.<GetMigrationCommandLists>b__2()
   at Microsoft.EntityFrameworkCore.Migrations.Internal.Migrator.Migrate(String targetMigration)
   at Microsoft.EntityFrameworkCore.RelationalDatabaseFacadeExtensions.Migrate(DatabaseFacade databaseFacade)
```

When I call EnsureCreated():

```
   at System.Linq.Enumerable.SingleOrDefault[TSource](IEnumerable`1 source, Func`2 predicate)
   at Microsoft.EntityFrameworkCore.Migrations.MigrationsSqlGenerator.FindProperty(IModel model, String schema, String tableName, String columnName)
   at Microsoft.EntityFrameworkCore.Migrations.MigrationsSqlGenerator.GetColumnType(String schema, String table, String name, Type clrType, Nullable`1 unicode, Nullable`1 maxLength, Boolean rowVersion, IModel model)
   at Microsoft.EntityFrameworkCore.Migrations.MigrationsSqlGenerator.ColumnDefinition(String schema, String table, String name, Type clrType, String type, Nullable`1 unicode, Nullable`1 maxLength, Boolean rowVersion, Boolean nullable, Object defaultValue, String defaultValueSql, String computedColumnSql, IAnnotatable annotatable, IModel model, MigrationCommandListBuilder builder)
   at Microsoft.EntityFrameworkCore.Migrations.SqliteMigrationsSqlGenerator.ColumnDefinition(String schema, String table, String name, Type clrType, String type, Nullable`1 unicode, Nullable`1 maxLength, Boolean rowVersion, Boolean nullable, Object defaultValue, String defaultValueSql, String computedColumnSql, IAnnotatable annotatable, IModel model, MigrationCommandListBuilder builder)
   at Microsoft.EntityFrameworkCore.Migrations.MigrationsSqlGenerator.ColumnDefinition(AddColumnOperation operation, IModel model, MigrationCommandListBuilder builder)
   at Microsoft.EntityFrameworkCore.Migrations.MigrationsSqlGenerator.Generate(CreateTableOperation operation, IModel model, MigrationCommandListBuilder builder)
   at Microsoft.EntityFrameworkCore.Migrations.SqliteMigrationsSqlGenerator.Generate(CreateTableOperation operation, IModel model, MigrationCommandListBuilder builder)
   at Microsoft.EntityFrameworkCore.Migrations.MigrationsSqlGenerator.<>c.<.cctor>b__53_8(MigrationsSqlGenerator g, MigrationOperation o, IModel m, MigrationCommandListBuilder b)
   at Microsoft.EntityFrameworkCore.Migrations.MigrationsSqlGenerator.Generate(MigrationOperation operation, IModel model, MigrationCommandListBuilder builder)
   at Microsoft.EntityFrameworkCore.Migrations.MigrationsSqlGenerator.Generate(IReadOnlyList`1 operations, IModel model)
   at Microsoft.EntityFrameworkCore.Migrations.SqliteMigrationsSqlGenerator.Generate(IReadOnlyList`1 operations, IModel model)
   at Microsoft.EntityFrameworkCore.Storage.RelationalDatabaseCreator.GetCreateTablesCommands()
   at Microsoft.EntityFrameworkCore.Storage.RelationalDatabaseCreator.CreateTables()
   at Microsoft.EntityFrameworkCore.Storage.RelationalDatabaseCreator.EnsureCreated()
   at Microsoft.EntityFrameworkCore.Infrastructure.DatabaseFacade.EnsureCreated()
```

I've tried deleteing and regenerating my migrations with the new tools. I'm still getting the error.

Strangely enough, my unit tests use InMemory database and they all run perfectly. It's only when I attempt this with Sqlite that it fails.

**Comment by ajcvickers** [comment]
@bricelam @rowanmiller @AndriySvyryd I sent out a fix for this but I haven't been able to come up with any workarounds in the meantime. Any ideas? It happens in Migrations whenever a column is mapped from more than one property.

**Comment by jakimber** [comment]
fwiw my Attempt 4 above works, and is what i have been using to keep developing.  Not a great one to release to the masses tho i'd imagine, but it may give ideas of another workaround to you guys.

Do i need to get nightly builds to get the fixes? or will there be a patch to the released feeds?

Thanks for this too!  The response is impressive!
